### PR TITLE
Created ReplaceableComponentsService and used it instead of ReplaceableComponentsState

### LINKED ISSUE #5026: Transform replaceable components from NGXS state to Angular service
Related to #4958

### PR DESCRIPTION
**Pull request by mehmet-erim**
Resolves #5026

### CODE REVIEW

File: npm/ng-packs/packages/core/src/lib/states/replaceable-components.state.ts [hunk 1]
```diff
@@ -69,6 +56,7 @@ export class ReplaceableComponentsState {
       replaceableComponents,
     });
 
-    if (reload) this.reloadRoute();
+    console.log(this.service);
```
Comment (armanozak) [inline npm/ng-packs/packages/core/src/lib/states/replaceable-components.state.ts]:
```suggestion
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import { Router } from '@angular/router';
+import { ReplaceableComponents } from '../models/replaceable-components';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { noop } from '../utils/common-utils';
+import { map, filter } from 'rxjs/operators';
+import { InternalStore } from '../utils/internal-store-utils';
+
+@Injectable({ providedIn: 'root' })
+export class ReplaceableComponentsService {
+  private state: InternalStore<ReplaceableComponents.ReplaceableComponent[]>;
```
Comment (armanozak) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
  private state = new InternalStore<ReplaceableComponents.ReplaceableComponent[]>([]);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  get replaceableComponents(): ReplaceableComponents.ReplaceableComponent[] {
+    return this.state.state;
+  }
+
+  get onUpdate$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
+    return this.state.sliceUpdate(state => state);
+  }
+
+  constructor(private ngZone: NgZone, private router: Router) {
+    this.state = new InternalStore([]);
```
Comment (armanozak) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts (hunk 2 above)
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]: Store would be better naming

```suggestion
  private store: InternalStore<ReplaceableComponents.ReplaceableComponent[]>;
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import { noop } from '../utils/common-utils';
+import { map, filter } from 'rxjs/operators';
+import { InternalStore } from '../utils/internal-store-utils';
+
+@Injectable({ providedIn: 'root' })
+export class ReplaceableComponentsService {
+  private state: InternalStore<ReplaceableComponents.ReplaceableComponent[]>;
+
+  get replaceableComponents$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
+    return this.state.sliceState(state => state);
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts:14]:
```suggestion
    return this.store.sliceState(state => state);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  get replaceableComponents(): ReplaceableComponents.ReplaceableComponent[] {
+    return this.state.state;
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts:18]:
```suggestion
    return this.store.state;
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  get onUpdate$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
+    return this.state.sliceUpdate(state => state);
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    return this.store.sliceUpdate(state => state);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts (hunk 3 above)
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    this.store = new InternalStore([]);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    this.router.navigated = false;
+
+    this.ngZone.run(async () => {
+      await this.router.navigateByUrl(this.router.url).catch(noop);
+      setRouteReuse(shouldReuseRoute);
+    });
+  }
+
+  add(replaceableComponent: ReplaceableComponents.ReplaceableComponent, reload?: boolean): void {
+    let replaceableComponents = this.state.state;
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    let replaceableComponents = this.store.state;
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      component => component.key === replaceableComponent.key,
+    );
+
+    if (index > -1) {
+      replaceableComponents[index] = replaceableComponent;
+    } else {
+      replaceableComponents = [...replaceableComponents, replaceableComponent];
+    }
+
+    this.state.patch(replaceableComponents);
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    this.store.patch(replaceableComponents);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import { BehaviorSubject, Observable } from 'rxjs';
+import { noop } from '../utils/common-utils';
+import { map, filter } from 'rxjs/operators';
+import { InternalStore } from '../utils/internal-store-utils';
+
+@Injectable({ providedIn: 'root' })
+export class ReplaceableComponentsService {
+  private store: InternalStore<ReplaceableComponents.ReplaceableComponent[]>;
+
+  get replaceableComponents$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts:13]:
```suggestion
  get replaceableComponents$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
     console.log('hello');
     return this.store.sliceState(state => state);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts (hunk 8 above)
Comment (armanozak) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    let found = 0;
    const replaceableComponents: ReplaceableComponents.ReplaceableComponent[] = [];

    this.state.state.forEach(component => {
      if (component.key === replaceableComponent.key) {
        replaceableComponents.push(replaceableComponent);
        found++;
        return;
      }

      replaceableComponents.push(component);
    });

    if (!found) replaceableComponents.push(replaceableComponent);

    this.state.patch(replaceableComponents);
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    const index = replaceableComponents.findIndex(
+      component => component.key === replaceableComponent.key,
+    );
+
+    if (index > -1) {
+      replaceableComponents[index] = replaceableComponent;
+    } else {
+      replaceableComponents = [...replaceableComponents, replaceableComponent];
+    }
```
Comment (armanozak) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
    const replaceableComponents = [...this.store.state];

    const index = replaceableComponents.findIndex(
      component => component.key === replaceableComponent.key,
    );

    if (index > -1) {
      replaceableComponents[index] = replaceableComponent;
    } else {
      replaceableComponents.push(replaceableComponent);
    }
```

File: npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+  get onUpdate$(): Observable<ReplaceableComponents.ReplaceableComponent[]> {
+    return this.store.sliceUpdate(state => state);
+  }
+
+  constructor(private ngZone: NgZone, private router: Router) {
+    this.store = new InternalStore([]);
+  }
+
+  // TODO: Create a shared service for route reload and more
```
Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts:30]:
```suggestion
```

Comment (bnymncoskuner) [inline npm/ng-packs/packages/core/src/lib/services/replaceable-components.service.ts]:
```suggestion
```

**Review by bnymncoskuner — APPROVED** [review]
LGTM

**Review by armanozak — APPROVED** [review]

**Comment by alexandis** [comment]
We have upgraded to ABP 3.1.2 and still cannot see ReplaceableComponentsService in @abp/ng.core, even though the documentation tells to use it for replacing the components. So do we still need to keep using old AddReplaceableComponent?

**Pull request merged**